### DISCUSSION TITLE
Pregunta Decisión de diseño TPI3

### DISCUSSION THREAD

**Discussion by ivanscardaoni1** [dds-jv/foro#300] [Consultas de objetos] [answered]
Quería preguntar una cuestión que me surgió sobre el TPI, pero que creo que aplica a varias situaciones de diseño.

   En este caso tengo que modelar los eventos repetitivos y los eventos únicos, pero quiero saber cuales son las ventajas y desventajas de tener una interfaz “Evento” de la que implementen “EventoRepetitivo” y “EventoUnico” que ambas respondan las ocurrencias de los eventos entre ciertas fechas como le corresponda a cada una, o tener una interfaz que sea “TipoDeEvento” de la que implementen “Repetitivo” y “Unico” y que eso sea un atributo de la clase “Evento” en la que se pueda delegar la resolución del método “Ocurrencias”.

   La pregunta más general seria eso, en base a que decido si es una referencia al objeto que lo resuelve y lo delego en el, o hago que la interfaz general ya marque la diferencia directamente. 

   Me imagino que un poco puede venir por el lado de cuanto necesitan conocer de la clase “Evento” las clases “TipoDeEvento”, pero como están acopladas si o si en esta situación no me queda muy claro porque haría una o la otra. 

   Me gustaría saberlo en términos generales para también en el parcial tener claro cómo justificar una u otra decisión.

**Comment by gastonprieto** [comment, marked as answer]
> La pregunta más general seria eso, en base a que decido si es una referencia al objeto que lo resuelve y lo delego en el, o hago que la interfaz general ya marque la diferencia directamente.

Mmm, es dificil responder a eso, porque no hay una única solución. Esta es un poco la diferencia entre composición y herencia (exceptuando que acá son 2 clases que implementan la misma interfaz), pero tenes el mismo problema.

- ¿Necesitas intercambiar su comportamiento?
- ¿Cuál solución es mas simple?
- ¿Cuál solución es mas extensible? ¿Necesitas esa extensibilidad?
- ¿Duplicas lógica?
- ¿Complejidad al instanciar?

La respuesta es depende, como el 99% de las cosas de la materia, perooo... es importante justificar porque se toma un camino, y eso también se va a evaluar, si en base al contexto es una justificación adecuada.
